### PR TITLE
[reland] Make grad point to bucket buffer in DDP to save memory usage

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44344 [reland] Make grad point to bucket buffer in DDP to save memory usage**

reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -316,6 +316,15 @@ class DistributedDataParallel(Module):
                          are getting different gradients, which should not
                          happen if DistributedDataParallel is correctly used.
                          (default: ``False``)
+        grad_is_view: when setting to ``True``, grads will be views of all reduce
+                      buckets in reducer, this can reduce peak memory usage, the
+                      saved memory size will be equal to the total grads size.
+                      Also it can avoid coping between grads and the all reduce
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:322]: > Also it can avoid coping between grads and the all reduce buckets in reducer and thus improve training performance.

-> Moreover, it avoids the overhead of copying between gradients and allreduce communication buckets.

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -316,6 +316,15 @@ class DistributedDataParallel(Module):
...
+                      buckets in reducer and thus improve training performance.
+                      When grads are views, detach_() can not be called in your
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:324]: grads -> gradients

double quote `detach_()`

> can not be called in your zero_grad() function, please

cannot be called on the gradients.  If hitting such errors, please

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -316,6 +316,15 @@ class DistributedDataParallel(Module):
...
+                      zero_grad() function, please fix it by referring to
+                      the Optimizer.zero_grad() function in torch/optim/optimizer.py
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:326]: double quote all code. 

BTW, can we make `zero_grad()` a link by using `:meth:torch.optim.Optimizer.zero_grad`?

**Review by mrshenli — COMMENTED** [review]
Lint errors are real. 

Could you please build the docs to verify the new doc strings are shown correctly? Thanks!

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -316,6 +316,15 @@ class DistributedDataParallel(Module):
                          are getting different gradients, which should not
                          happen if DistributedDataParallel is correctly used.
                          (default: ``False``)
+        grad_is_view: when setting to ``True``, grads will be views of all reduce
```
Comment (pritamdamania87) [inline torch/nn/parallel/distributed.py:319]: nit: `when set to True` instead of `when setting to True`.

File: test/distributed/test_c10d.py [hunk 5]
```diff
@@ -2749,6 +2748,16 @@ def step_model(model, input, target):
...
 
+    @requires_nccl()
+    @skip_if_not_multigpu
+    def test_accumulate_gradients_module(self):
+        self._test_accumulate_gradients_module()
+
+    @requires_nccl()
+    @skip_if_not_multigpu
+    def test_accumulate_gradients_module_with_grad_is_view(self):
+        self._test_accumulate_gradients_module(grad_is_view=True)
```
Comment (pritamdamania87) [inline test/distributed/test_c10d.py:2759]: Can we add a test where grad_is_view is True and we're using ddp_comm_hook?

File: test/distributed/test_c10d.py [hunk 6]
```diff
@@ -1973,13 +1973,15 @@ def tearDown(self):
     def world_size(self):
         return 2
 
-    def _prepare_single_device_module(self, process_group, devices, device_ids, global_batch_size):
+    def _prepare_single_device_module(
```
Comment (mrshenli) [inline test/distributed/test_c10d.py:1976]: Shall we turn on more tests? like:

1. `test_multiple_outputs_multiple_backward`
2. `test_accumulate_gradients_no_sync`
3. `test_arbitrary_forward_return_value`
4. etc.

Will they work under this mode?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 7]
```diff
@@ -1015,24 +1104,53 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
...
-          return true;
+      if (!gradient_as_bucket_view_) {
+        copy_bucket_to_grad(
+            variable, replica, intra_bucket_index, global_unused);
+      } else {
+        const auto& bucket_view_out =
+            replica.bucket_views_out[intra_bucket_index];
+        auto& bucket_view_in = replica.bucket_views_in[intra_bucket_index];
+        // If communication_hook is registered, bucket_view_out stores
+        // all reduced results in a newly allocated tensor, copy bucket_view_out
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:1115]: all reduced -> allreduced

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 8]
```diff
@@ -1015,24 +1104,53 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
...
-      });
+        runGradCallbackForVariable(variable, [&](auto& grad) {
+          // If a parameter is globally unused, we keep its grad untouched.
+          if (!global_unused) {
+            // If grad is globally used but locally unused, let grad point to
+            // bucket_view_in
+            if (!grad.defined()) {
+              grad = bucket_view_in;
+            } else {
+              if (!grad.is_alias_of(bucket_view_in)) {
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:1129]: do we also need to guard this with `gradient_as_bucket_view_` flag, as we shouldn't print this warning when this mode is off?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 9]
```diff
@@ -1015,24 +1104,53 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
...
+                    "the same storage. For example, param0 and param1 "
+                    "share the same gradient variable grad0. grad0 "
+                    "firstly points to bucket_view_in0 when param0 gradient "
+                    "is ready, then grad0 points to bucket_view_in1 when "
+                    "param1 gradient is ready during  "
+                    "''mark_variable_ready_dense''. In this case, grad0 "
+                    "is not alias of bucket_view_in0 here, although it can "
+                    "still copy bucket_view_in0 to grad0. If you have this "
+                    "case, please check whether it is expected in your "
+                    "application.");
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:1144]: Let's reword a bit:

Detected at least one parameter gradient is not the expected DDP bucket view when setting `gradient_as_bucket_view=True`. This can happen when multiple parameters sharing the same gradient. For example, param0 and param1 share the same gradient grad0. In this case, grad0 would first point to `bucket_view_in0` when param0 is ready. Later, when param1 is ready, it will override grad0 to point to `bucket_view_in1`. However, param0 still expects grad0 to point to `bucket_view_in0`, and hence hit this warning. If you saw this message, please double-check if the above situation is expected for your application.

File: torch/csrc/distributed/c10d/reducer.h [hunk 10]
```diff
@@ -230,6 +232,19 @@ class Reducer {
   // with the result of `future_work`.
   void populate_bucket_views_out(BucketReplica& replica, at::Tensor& tensor);
 
+  // If gradient_as_bucket_view_ is false, after all reduce buckets,
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.h:235]: all reduce -> allreduce

File: torch/csrc/distributed/c10d/reducer.h [hunk 11]
```diff
@@ -230,6 +232,19 @@ class Reducer {
...
+  // copy bucket results back to grads.
+  void copy_bucket_to_grad(
+      torch::autograd::Variable& variable,
+      Reducer::BucketReplica& replica,
+      size_t intra_bucket_index,
+      bool global_unused);
+  // Check layout of grad and bucket_view before calling copy_grad_to_bucket
+  void check_grad_layout(const at::Tensor& grad, const at::Tensor& bucket_view);
+  // If gradient_as_bucket_view_ is false, before all reduce buckets,
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.h:244]: ditto

File: torch/nn/parallel/distributed.py [hunk 12]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
                          are getting different gradients, which should not
                          happen if DistributedDataParallel is correctly used.
                          (default: ``False``)
+        gradient_as_bucket_view: this is a prototype feature. When set to ``True``,
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:318]: `gradient_as_bucket_view` -> `gradient_as_bucket_view (bool)`

File: torch/nn/parallel/distributed.py [hunk 13]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      gradients will be views pointing to different offsets of
+                      allreduce communication buckets. This can reduce peak memory
+                      usage, where the saved memory size will be equal to the total
+                      grads size. Moreover, it avoids the overhead of copying
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:322]: grads -> gradients

File: torch/nn/parallel/distributed.py [hunk 14]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      between gradients and allreduce communication buckets.
+                      When gradients are views, "detach_()" cannot be called on the
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:324]: by double quote, I was actually referring to double italic quote, i.e. 
```
``code``
```
This will make our docgen to recognize it as a code.

File: torch/nn/parallel/distributed.py [hunk 15]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      gradients. If hitting such errors, please fix it by referring to
+                      the :meth:torch.optim.Optimizer.zero_grad function in
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:326]: I think this should be.

```
:meth:`~torch.optim.Optimizer.zero_grad`
```

File: torch/nn/parallel/distributed.py [hunk 16]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      "torch/optim/optimizer.py" as the solution.
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:327]:
```
``torch/optim/optimizer.py``
```

File: torch/nn/parallel/distributed.py [hunk 17]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      Warning! It is also found that 'gradient_as_bucket_view = true'
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:328]: warnings are placed in separate blocks. Please search for `.. warning::` in this file.

Comment (mrshenli) [inline torch/nn/parallel/distributed.py:328]:
```
``gradient_as_bucket_view = true``
```

File: torch/nn/parallel/distributed.py [hunk 18]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      does not work as expected when apex.amp is used for
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:329]:
```
``apex.amp``
```

File: torch/nn/parallel/distributed.py [hunk 19]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      mixed precision training. apex.amp maintained stashed gradients
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:330]:
```
``apex.amp``
```

File: torch/nn/parallel/distributed.py [hunk 20]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      that are used for unscaling gradients. These stashed gradients
+                      are pointed to gradients (will be communication buckets when
+                      'gradient_as_bucket_view = true') before starting new iteration.
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:333]:
```
``gradient_as_bucket_view = true``
```

File: torch/nn/parallel/distributed.py [hunk 21]
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
...
+                      In new iteration, the communication buckets are mutated and thus
+                      these stashed gradients will be unexpectedly mutated as well,
+                      the unexpectedly muated stached gradients may result in wrong
+                      results. To fix it, these stashed gradients should not be pointed
+                      to gradients, instead they should be copied from gradients when
+                      'gradient_as_bucket_view = true'.
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:339]:
```
``gradient_as_bucket_view = true``
```

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 8 above)
Comment (zhaojuanmao) [reply torch/csrc/distributed/c10d/reducer.cpp:1129]: when this mode is off, it will go through 'copy_bucket_to_grad' code path, this is only for gradients_as_bucket_view=true

File: torch/nn/parallel/distributed.py (hunk 15 above)
Comment (mrshenli) [reply torch/nn/parallel/distributed.py:326]: I might be wrong, does `:meth:torch.optim.Optimizer.zero_grad` also correctly show as a link?

**Review by mrshenli — APPROVED** [review]
LGTM!

Master was broken and is fixed now. Could you please rebase to the latest commit and wait for all tests to pass before landing? Thanks!

**Comment by zhaojuanmao** [comment]
failures are not related

**Comment by zhaojuanmao** [comment]
merge conflict, rebase
